### DISCUSSION TITLE
Virgl not functionning under Apple's virtualization.framework

### DISCUSSION THREAD

**Discussion by ZeppLu** [utmapp/UTM#5482] [Linux] [answered]
I created two VMs, both of which are Ubuntu 22.04, under virtualization framework and qemu respectively.

`lspci -v` in both VMs show the presence of virtio GPU, but only the qemu one reports virtio_gpu as OpenGL driver in `glxinfo -B`, which the apple one reports llvmpipe. Has anyone gotten the same issue? Is such behaviour expected?

**Comment by osy** [comment, marked as answer]
There is no GPU acceleration under AVF

**Reply by ZeppLu** [reply]
May you explain a little more about what prevents such feature?

AFAIK, renderding under qemu is accelerated by passing virgl to libvirglrenderer.so (correct me if I'm wrong). So AVF just ignores virgl commands and renders everything internally, without exposing customizable options?

**Reply by osy** [reply]
What prevents such a feature is that apple hasn't implemented it. There is no QEMU if you choose virtualization.framework.

**Comment by ryuheechul** [comment]
_elaborating a bit more on https://github.com/utmapp/UTM/discussions/5482#discussioncomment-6502033 and sharing what I recently learned on the subject_

You can see that apple only added virtio gpu 2d and no further anouncement on that.
> [In macOS Ventura, we have added support for Virtio GPU 2D](https://developer.apple.com/videos/play/wwdc2022/10002?time=1077)

There is 3d acceleration via qemu (not with the virtualization framework) which is pretty satisafactory for my usage with [alacritty](https://github.com/alacritty/alacritty).

Currently my exeprience on graphics with few VM guests:
- linux guest (nixos) qemu via [hvf](https://developer.apple.com/documentation/hypervisor)
  - great with overall UIs
  - video playback (via youtube) is watchable but audio sync with video comes and goes
- linux guest (nixos) with virtualization framework
  - no 3d acceleartion (so GUI stuff is laggy in general including alacritty)
  - video playback is good except the frame rate (but audio sync is pretty good)
  - (not sure if it's related to graphics) but freezing occurs a lot
    - not just with utm but also with other apps that uses virtualization framework 
- macOS guest with virtualization framework
  - graphical performance is really good (pretty much can't feel the difference with the host except it 60hz seems to be the limit, I couldn't find an option to enable 120hz)
    - also the score for https://www.wirple.com/bmark/ was similar to the one from the host
  
I think for the virtualization framework to compete with QEMU (with hvf) (in temrs of GUI usages), apple need to start supporting 3d acceleration for linux guests.

**Reply by JezerM** [reply]
Hi, how are you able to run alacritty in a Linux guest via qemu? No matter the backend (OpenGL nor Metal), alacritty doesn't render anything on screen.
